core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -526,6 +534,45 @@ private boolean expandSelectItem(final SqlNode selectItem, SqlSelect select,
     return false;
   }
 
+  /**
+   * Returns true if the node is a non-aggregated, non-grouped column in SELECT.
+   */
+  private boolean isNonAggregatedNonGroupedColumn(SqlNode node, SqlSelect select) {
+    if (aggFinder.findAgg(node) != null) {
+      return false;
+    }
+
+    if (node instanceof SqlIdentifier) {
+      SqlNodeList groupList = select.getGroup();
+      if (groupList == null) {
+        return true;
+      }
+      return groupList.getList().stream()
+          .noneMatch(groupItem -> groupItem != null
+              && equalAsIdentifier((SqlIdentifier) node, groupItem));
+    }
+
+    if (node instanceof SqlCall) {
+      return ((SqlCall) node).getOperandList().stream()
+          .allMatch(operand -> isNonAggregatedNonGroupedColumn(operand, select));
+    } else if (node instanceof SqlLiteral) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /** Determines whether two SqlIdentifier objects are equivalent. */
+  private boolean equalAsIdentifier(SqlIdentifier id1, SqlNode node2) {
+    if (!(node2 instanceof SqlIdentifier)) {
```

Comment:
This function needs many more tests. 
Ideally you would not compare identifiers, but column indexes in inputs.
Are you sure that the compared identifiers are both fully expanded at this point?